File: src/solutions/2023/09/lua/main.lua
local run = require('src.langs.lua.run')
local str = require('src.langs.lua.utils.str')

local options = {
   has_alternate = false,
   has_io = false,
}

local function difference(ary, sz)
   for i = 1, sz do
      ary[i] = ary[i + 1] - ary[i]
   end
   return ary
end

local function extrapolate(ary, n)
   n = n - 1
   local last = ary[n + 1]
   if n == 1 then
      return last
   end
   return extrapolate(difference(ary, n), n) + last
end

---@param input string
---@param is_test boolean
---@return string
local function part_1(input, is_test)
   local lines = str.split(input, '\n')
   
   local res = 0
   for i = 1, #lines do
      local line = str.split(lines[i], ' ')
      local nums = {}

      for j = 1, #line do
         nums[#nums + 1] = tonumber(line[j])
      end
      res = res + extrapolate(nums, #nums)
   end

   return tostring(res)
end

---@param input string
---@param is_test boolean
---@return string
local function part_2(input, is_test)
   local lines = str.split(input, '\n')
   
   local res = 0
   for i = 1, #lines do
      local line = str.split(lines[i], ' ')
      local nums = {}

      for j = 1, #line do
         nums[#nums + 1] = tonumber(line[#line + 1 - j])
      end
      res = res + extrapolate(nums, #nums)
   end

   return tostring(res)
end

local function main()
   run(arg, part_1, part_2, options)
end

main()

File: src/solutions/2023/09/rust/main.rs
extern crate aoc_core;
use std::{env, iter::FromIterator};

fn get_options() -> aoc_core::options::SolutionOptions {
   aoc_core::options::SolutionOptions {
      ..Default::default()
   }
}

fn difference(ary: &mut Vec<i64>, sz: usize) -> &mut Vec<i64> {
   for i in 0..sz {
      ary[i] = ary[i + 1] - ary[i];
   }
   ary
}

fn extrapolate(ary: &mut Vec<i64>, sz: usize) -> i64 {
   let z = sz - 1;
   let last = ary[z];
   if z == 0 {
      last
   } else {
      extrapolate(difference(ary, z), z) + last
   }
}

fn part_1(input: &str, _is_test: bool) -> String {
   let parsed: Vec<Vec<i64>> = input
      .lines()
      .map(|x| x.split(' ').map(|y| y.parse::<i64>().unwrap()).collect())
      .collect();
   let mut res = 0_i64;
   for mut history in parsed {
      let len = history.len();
      res += extrapolate(&mut history, len);
   }
   res.to_string()
}

fn part_2(input: &str, _is_test: bool) -> String {
   let parsed: Vec<Vec<i64>> = input
      .lines()
      .map(|x| x.split(' ').map(|y| y.parse::<i64>().unwrap()).rev().collect())
      .collect();
   let mut res = 0_i64;
   for mut history in parsed {
      let len = history.len();
      res += extrapolate(&mut history, len);
   }
   res.to_string()
}

fn main() {
   let args = Vec::from_iter(env::args());
   aoc_core::run(
      args.iter().map(|x| x.as_ref()).collect::<Vec<_>>(),
      part_1,
      part_2,
      get_options(),
   );
}
